### ISSUE TITLE
What should "nested" dialog light dismiss look like?

### ISSUE THREAD

**Issue by mfreed7** [openui/open-ui#1128]
I have a spec PR open to add dialog "light dismiss" capabilities, via a new `closedby` attribute, here: https://github.com/whatwg/html/issues/9373. One question came up (see discussion starting [roughly here](https://github.com/whatwg/html/issues/9373#issuecomment-2490229365)) about what the behavior should be for "nested" dialogs. I keep putting quotes around "nested" because dialogs don't have any current notion of being "nested", in the same way that popovers explicitly do (see the big `Note` here: https://html.spec.whatwg.org/#topmost-popover-ancestor). So any behavior we add here is essentially inventing some concept of nested dialogs.

So the question is, what are typical expectations for users or developers when multiple modal dialogs are open, and the user clicks outside one of them. I'd like to side-step the question of modeless dialogs, at least for now, since I consider modals the more pressing, and common, use case.

[Here's a demo to work with](https://jsbin.com/pixuxin/edit?html,output) (borrowed and tweaked from @domenic):

![Screenshot 2024-11-22 at 10 15 25 AM](https://github.com/user-attachments/assets/727f614f-26c2-4ac3-b381-2db03d27c0aa)

Some things to note:
 - while the dialogs definitely look "nested", they are in fact siblings in the DOM tree. Their order in the DOM does not matter for this visual presentation, only the order of the `showModal` calls.
 - both dialogs are modal, so both have a `::backdrop` covering the area outside the dialog.
 - due to the above, a click outside `inner` is actually a click on `inner`'s `::backdrop`. It is impossible to click on `outer` or its `::backdrop`.

There are three positions that a user might try to click (labeled in red in the demo above):
 1. Within `inner`'s border box.
 2. Within the visible portion of `outer`'s border box.
 3. Outside both dialogs' border boxes.

Note, again, that <span>#</span>2 and <span>#</span>3 are both actually clicks on the backdrop of `inner`.

In terms of potential outcomes, for each click position:
1. Clicking on `inner` should always leave `inner` visible. It could either close `outer` or leave it open. It seems like it makes the most sense to leave `outer` open, at least to a user for this use case.
2. This is a click outside `inner`, so `inner` should clearly close. And because it's an attempted click on `outer`, it seems like `outer` should stay open.
3. Clearly, `inner` should close, since it is topmost. The question is whether the user expects this click to just close `inner` and leave `outer` open (i.e. just pop one dialog off the stack), or whether both should close. My assumption is that in all design systems today that implement "light dismiss dialogs", `outer` stays open, just due to implementation. But I'd love confirmation.

The point of this issue is a) to make sure I haven't missed something, and b) to check expectations for behavior. I'd love to discuss, so Agenda+.

**Comment by keithamus** [comment]
> 1. Clicking on `inner` should always leave `inner` visible. It could either close `outer` or leave it open. It seems like it makes the most sense to leave `outer` open, at least to a user for this use case.

Yes, it's important that clicking on inner should not close any lower dialogs. It should leave `outer` open.
 
> 2. This is a click outside `inner`, so `inner` should clearly close. And because it's an attempted click on `outer`, it seems like `outer` should stay open.

Right. Only one dialog should close at a time. Closing many may be an unpredictable behaviour, and would sow distrust for the user as to what they can anticipate.

> 3. Clearly, `inner` should close, since it is topmost. The question is whether the user expects this click to just close `inner` and leave `outer` open (i.e. just pop one dialog off the stack), or whether both should close. My assumption is that in all design systems today that implement "light dismiss dialogs", `outer` stays open, just due to implementation. But I'd love confirmation.

We have some nested dialogs where `outer` stays open, and it is a carefully considered design choice. It is important to us that the implied behaviour is "just pop one off the stack" because it allows for a very predictable consistency, whether or not the developer made them nested or not, and regardless of the "type" (hint, popover, dialog, etc). 

![a screenshot of the "Create new issue" dialog in GitHub, where some changes have been made, and the user has tried to dismiss the dialog. A nested dialog lays over the top with the heading "Discard changes?"](https://github.com/user-attachments/assets/b2444cb8-8e24-4150-b04d-803859ebeca8)

In the above image, there is actually a more complex case: the `Discard changes?` dialog can be light dismissed, which acts like the `Keep editing` button. Attempting to light dismiss dialog "Create new issue" will _invoke_ the "Discard changes?" dialog, in order to intercept the user before they are able to dismiss this dialog and potentially reset their work. For us it is important that clicking the backdrop always does something; either light dismiss the current top dialog, or present a new dialog (or some kind of visual confirmation) showing that the light dismiss has been prevented.

**Comment by mfreed7** [comment]
> We have some nested dialogs where `outer` stays open, and it is a carefully considered design choice. It is important to us that the implied behaviour is "just pop one off the stack" because it allows for a very predictable consistency, whether or not the developer made them nested or not, and regardless of the "type" (hint, popover, dialog, etc).

This is very helpful - thanks.

> In the above image, there is actually a more complex case: the `Discard changes?` dialog can be light dismissed, which acts like the `Keep editing` button. Attempting to light dismiss dialog "Create new issue" will _invoke_ the "Discard changes?" dialog, in order to intercept the user before they are able to dismiss this dialog and potentially reset their work. For us it is important that clicking the backdrop always does something; either light dismiss the current top dialog, or present a new dialog (or some kind of visual confirmation) showing that the light dismiss has been prevented.

Thanks also for this use case. I believe this should be easy to implement, within the currently proposed behavior (https://github.com/whatwg/html/issues/9373), by adding a listener for the outer dialog's `cancel` event, and `preventDefault`ing it as needed. Does that seem correct?

**Comment by keithamus** [comment]
> Thanks also for this use case. I believe this should be easy to implement, within the currently proposed behavior ([whatwg/html#9373](https://github.com/whatwg/html/issues/9373)), by adding a listener for the outer dialog's `cancel` event, and `preventDefault`ing it as needed. Does that seem correct?

Yes absolutely. No concerns on that particular behaviour.

**Comment by mfreed7** [comment]
So I updated Chrome's behavior to match what's roughly being proposed in the conversation above, namely that clicking outside a "stack" of dialogs only "pops" the first one off the stack. So for N dialogs, you'll need N clicks outside to close them all. The behavior for popovers meshes fairly well with this also, or at least that was my impression. So I've roughly come around to thinking this is likely the better/correct behavior.

But if folks want to try it out live, grab the next Chrome Canary release (see [this link for the exact version](https://chromiumdash.appspot.com/commit/0aadd6e7f5dd977823c4ec3715573ecdc91e1281)) and give it a try! Please remember to add `<dialog closedby=any>` to your dialog, or you'll get the existing behavior - no light dismiss.

**Comment by lukewarlow** [comment]
Do you have an actual demo for this (sorry to be lazy) just so I can see what this kind of nesting looks like. But to me this sounds like what I would intuit as the behaviour.

**Comment by sorvell** [comment]
>  just pop one dialog off the stack

I definitely think this is the correct behavior. 

Sorry this is probably a bit obtuse, but in general a modal dialog is intended to / must be responded to in some way.  I think it's common in these cases that a "parent" dialog spawns the "nested" one, and this often obscures the "parent." Responding to the "nested" dialog does *not* respond to the "parent," and as a user I expect to be able to get back to the "parent" and respond to it directly. Otherwise, I'm flummoxed with "where did that other dialog go?"

**Comment by lukewarlow** [comment]
The update definitely feels like the better behaviour.

**Comment by lukewarlow** [comment]
https://jsbin.com/pesomokilo/edit?html,output - The behaviour when you press escape here closing the back modal feels perhaps wrong to me (I could be convinced either way though)? Is this behaving as specced or is this expected to change?

**Comment by domenic** [comment]
> The behaviour when you press escape here closing the back modal feels perhaps wrong to me (I could be convinced either way though)? Is this behaving as specced or is this expected to change?

This is due to the current spec's CloseWatcher anti-abuse grouping. A more realistic example that has user activation does not exhibit the problem: https://jsbin.com/yefajififu/edit?html,output

**Comment by keithamus** [comment]
@domenic do you think the internals of `showModal()` could circumvent the anti-abuse grouping? If a web author has called `showModal()` on a lot of dialogs then closing them in an order different to what they're shown does seem odd. It is of course an edge case but it might be tractable to solve for, no?

**Comment by domenic** [comment]
I'm not clear on what you're envisioning. Circumventing anti-abuse grouping sounds like allowing abuse, which is bad?

**Comment by domenic** [comment]
That trivially allows abuse: you then call `showModal()` N times and you now have trapped the user (they require N back presses to exit the site).

The point of the anti-abuse mechanisms is to cap the number of back presses at (number of user activations) + 1, i.e. put the _user_ in control. Capping it at (number of `showModal()` calls) + 1 puts the developer in control, but the developer is the potential abuser.

https://github.com/WICG/close-watcher?tab=readme-ov-file#abuse-analysis may be worth reviewing.

**Comment by lukewarlow** [comment]
Fwiw I'm fine with the odd behaviour given it only applies in these anti-abuse cases, and with the mix of closedby values. It feels edge case enough that it'll almost never come up (other than for us spec and implementor people who think about these cases)

**Comment by mfreed7** [comment]
Pending @keithamus's agreement, it sounds like perhaps we've somewhat arrived at consensus? For me at least, with the new behavior, both [this](https://jsbin.com/pesomokilo/edit?html,output) (@lukewarlow's example) and [this](https://jsbin.com/yefajififu/edit?html,output) (@domenic's example) feel "ok" to me. Both are somewhat odd situations, where the "outer" dialog is `closedby=any` but the inner is `closedby=none`. I would think that in a real situation, if your outer dialog can open non-light-dismissible dialogs, then it shouldn't be light dismissible itself. Or, if you need that, use @domenic's code path, where the inner one is user activated.

Assuming there are no other objections, I'm inclined to go back to [the spec PR](https://github.com/whatwg/html/pull/10737) and say we now have the right behavior. So... any objections to that?

**Comment by domenic** [comment]
So I think I finally understand what the last few comments were getting at. They are objecting that **light dismiss** is blocked by the anti-abuse mechanisms, even though the anti-abuse mechanisms are intended to only apply to **close requests**.

This happens in the current spec PR (and presumably Chrome implementation) because @mfreed7 is using "request to close _topmostDialog_'s close watcher" as the final step of "light dismiss open dialogs".

I am unsure what path is best.

- `requestClose()` on `CloseWatcher`s, and (per the proposed spec) `<dialog>`s, is intended to imitate a close request. This includes the anti-abuse behavior.

- But the anti-abuse behavior is not necessary for light dismiss, nor is it necessary for `requestClose()`. And it can be confusing for developers, as seen above.

Should we instead have light dismiss go down an alternate path, which does not include the anti-abuse behavior? What about `requestClose()`?

**Comment by domenic** [comment]
(The above post is a bit confused. I will hide it. This is a second attempt.)

Although the above comments are talking about the behavior for close requests (Esc key / Android back button), there is a related issue which is whether the anti-abuse mechanisms should apply to light dismiss and `requestClose()`.

Recall that the anti-abuse protections come in two parts:

1. Inside [request to close](https://html.spec.whatwg.org/#close-watcher-request-close), we only allow the `cancel` event to be `preventDefault()`ed under certain circumstances, and we consume history-action user activation if `preventDefault()` is called.
2. Inside [process close watchers](https://html.spec.whatwg.org/#process-close-watchers), we can close multiple close watchers at the same time if they are grouped. (Grouping happens when you create multiple close watchers without user activation.)

(2) is only relevant for close requests, so that's fine. Most of the above discussions were talking about (2).

But (1) is being applied to `CloseWatcher`'s `requestClose()`, and (per the proposal in https://github.com/whatwg/html/pull/10737) to `<dialog>` light dismiss and `requestClose()`. It does not have to be, since there is no abuse potential in those scenarios.

Should we consider bypassing (1) for those cases? To spec this I would probably add an optional _requireHistoryActionActivation_ boolean to "request to close" and set it to true for close requests, and false for light dismiss and `requestClose()`. This might require updating some test expectations for `CloseWatcher`.

**Comment by mfreed7** [comment]
> But (1) is being applied to `CloseWatcher`'s `requestClose()`, and (per the proposal in [whatwg/html#10737](https://github.com/whatwg/html/pull/10737)) to `<dialog>` light dismiss and `requestClose()`. It does not have to be, since there is no abuse potential in those scenarios.
> 
> Should we consider bypassing (1) for those cases? To spec this I would probably add an optional _requireHistoryActionActivation_ boolean to "request to close" and set it to true for close requests, and false for light dismiss and `requestClose()`. This might require updating some test expectations for `CloseWatcher`.

Ohh this is a good point. I agree that there doesn't seem to be any abuse potential in either light dismiss (where the user isn't hitting a "close" button which might be trying to navigate back) or in the developer calling `dialog.requestClose()` (the user isn't doing anything directly). So I'm in favor of this change.

Any objections from developers here? I don't see why there would be objections, but just checking. Otherwise, I'll take a look at adding some way around the abuse protections in the [dialog spec PR](https://github.com/whatwg/html/pull/10737). @domenic, if you have specific suggestions for a good way to go about that, I'm all ears. My first thought is to just add a boolean parameter to `request to close` that will force `canPreventClose` to true.

**Comment by mfreed7** [comment]
> Otherwise, I'll take a look at adding some way around the abuse protections in the [dialog spec PR](https://github.com/whatwg/html/pull/10737). @domenic, if you have specific suggestions for a good way to go about that, I'm all ears. My first thought is to just add a boolean parameter to `request to close` that will force `canPreventClose` to true.

https://github.com/whatwg/html/pull/10737/commits/6aa73868d7cbf24a35f580ac703a77bdd8315bde

**Comment by mfreed7** [comment]
> Should we consider bypassing (1) for those cases? To spec this I would probably add an optional _requireHistoryActionActivation_ boolean to "request to close" and set it to true for close requests, and false for light dismiss and `requestClose()`. This might require updating some test expectations for `CloseWatcher`.

@domenic somehow I totally missed this paragraph of your comment. Thanks for the specific suggestion - I [just flipped](https://github.com/whatwg/html/pull/10737/commits/344d87fcd6d95418e5995e34a4a352b3abdf3c91) the name and logic of my last commit to match. Note that I think we should handle the `closeWatcher.requestClose()` case separately because it might have compat issues. I doubt it, but I'd like to separate it from the dialog stuff, which doesn't have that problem.

**Comment by domenic** [comment]
> Note that I think we should handle the `closeWatcher.requestClose()` case separately because it might have compat issues. I doubt it, but I'd like to separate it from the dialog stuff, which doesn't have that problem.

I'm a little sad about introducing `dialogEl.requestClose()` which has different behavior than `closeWatcher.requestClose()`, so my instinct would be to bundle that change in. Especially given how unlikely the compat issues would be. (It's relatively hard to depend on an event being non-`cancelable`, I think...)

But, I trust you to fast-follow with such a change, if you would prefer to split them.

**Comment by mfreed7** [comment]
> I'm a little sad about introducing `dialogEl.requestClose()` which has different behavior than `closeWatcher.requestClose()`, so my instinct would be to bundle that change in. Especially given how unlikely the compat issues would be. (It's relatively hard to depend on an event being non-`cancelable`, I think...)

Yeah that's fair, compat might not be that bad. Ok. I'll try to put them together and ship them both at the same time. 😄

**Comment by mfreed7** [comment]
I'm going to close this issue, since we discussed and resolved on the latest spec behavior.
